nx/lib/nx/random.ex

```diff
@@ -0,0 +1,390 @@
+      |> Nx.bitwise_xor(y1)
+
+    Nx.stack([y1, y2])
+    |> Nx.as_type({:u, 32})
+  end
+
+  defnp rolled_loop_step(i, {xs, ks, rs}) do
+    {k1, k2, k3} = {ks[0], ks[1], ks[2]}
+    {r1, r2} = {rs[0], rs[1]}
+
+    {_, xs, _} =
+      while {x = 0, xs, rs}, Nx.less(x, 4) do
+        {x + 1, apply_round(xs, rs[0][x]), rs}
+      end
+
+    xs1 =
+      Nx.broadcast(k1, xs[0])
+      |> Nx.add(xs[0])
+
+    xs2 =
+      Nx.broadcast(k2 + i + 1, xs[1])
+      |> Nx.add(xs[1])
+
+    new_xs =
+      Nx.stack([xs1, xs2])
+      |> Nx.as_type({:u, 32})
+
+    new_ks = Nx.stack([k2, k3, k1])
+    new_rs = Nx.stack([r2, r1])
+
+    {new_xs, new_ks, new_rs}
+  end
+
+  defnp rotate_left(x, rot) do
+    nbits = 32
+    x <<< rot ||| x >>> (nbits - rot)
+  end
+
+  defp finfo(type) do
+    case type do
+      {:bf, 16} -> [exp: 8, mant: 7]
+      {:f, 16} -> [exp: 5, mant: 11]
+      {:f, 32} -> [exp: 8, mant: 23]
+      {:f, 64} -> [exp: 11, mant: 52]
+    end
+  end
+
+  @doc """
+  Sample uniform random integer values in [minval, maxval).
+  
+  ## Options
+  
+    * `:type` - an int type for the returned tensor
+  
+    * `:shape` - shape of the returned tensor
+  
+  """
+  defn randint(key, minval, maxval, opts \\ []) do
+    opts = keyword!(opts, shape: {1}, type: {:s, 32})
+    shape = opts[:shape]
+    type = {_, nbits} = opts[:type]
+
+    transform(key, &randint_transform(&1, type))
+
+    keys = transform(key, &split/1)
+
+    minval = Nx.broadcast(minval, shape)
+    maxval = Nx.broadcast(maxval, shape)
+
+    higher_bits = transform(keys[0], &random_bits(&1, shape))
+    lower_bits = transform(keys[1], &random_bits(&1, shape))
+    span = maxval - minval
+
+    multiplier =
+      Nx.power(2, Nx.quotient(nbits, 2))
+      |> Nx.remainder(span)
+      |> Nx.power(2)
+      |> Nx.remainder(span)
+
+    offset =
+      higher_bits
+      |> Nx.remainder(span)
+      |> Nx.multiply(multiplier)
+      |> Nx.add(Nx.remainder(lower_bits, span))
+      |> Nx.remainder(span)
+
+    Nx.as_type(minval + offset, type)
+  end
+
+  defp randint_transform(key, type) do
+    assert_key(key)
+
+    if not Nx.Type.integer?(type) do
+      raise ArgumentError,
+            "expected integer type, got type #{inspect(type)}"
+    end
+  end
+
+  @doc """
+  Sample uniform float values in [minval, maxval).
+  
+  ## Options
+  
+    * `:type` - an int type for the returned tensor
+  
+    * `:shape` - shape of the returned tensor
+  
+    * `:minval` - minimum value, default is 0.0
+  
+    * `:maxval` - maximum value, default is 1.0
+  
+  """
+  defn uniform(key, opts \\ []) do
+    opts = keyword!(opts, shape: {1}, type: {:f, 32}, minval: 0.0, maxval: 1.0)
+
+    transform(key, &uniform_transform(&1, opts))
+
+    info = transform(opts[:type], &finfo(&1))
+
+    shape = opts[:shape]
+    type = {_dtype, nbits} = opts[:type]
+    minval = opts[:minval]
+    maxval = opts[:maxval]
+
+    u_one = Nx.tensor(1.0, type: type) |> Nx.bitcast({:u, nbits})

```

Comment:
Is there any way to avoid bitcast here?
We can't have bitcast in Torchx.

Since it's the beggining of the function, I would rather do the bitcast in a transform in pure elixir if it is indeed necessary.

Reply:
u_one isn't actually one, right? What is the idea here?

Reply:
> u_one isn't actually one, right? What is the idea here?

It let's us preserve the exponent of 1 (we are only randomizing fraction/mantissa)